src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -278,31 +281,99 @@ def process_input(conf):
         if not task:
             print("Stopping")
             break
-        upa, obj_info, genome_upa = task
+        upa, obj_info, genome_upa, retrieve_sample = task
+
+        upa_dir = os.path.join(conf.pth, upa)
+        if not os.path.isdir(upa_dir) or not loader_helper.is_upa_info_complete(upa_dir):
+
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            if os.path.isdir(upa_dir):
+                shutil.rmtree(upa_dir)
+
+            # cfn points to the assembly file outside of the container
+            # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
+            # workdir is shared between the container and the external file system
+            # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+            cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+            # upa file is downloaded to cfn
+            conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+            # each upa in output_dir as a separate directory
+            dstd = os.path.join(conf.pth, upa)
+            os.makedirs(dstd, exist_ok=True)
+
+            dst = os.path.join(dstd, f"{upa}.fa")
+            # Hard link .fa file from job_dir to output_dir in WS
+            os.link(cfn, dst)
+
+            metafile = os.path.join(dstd, f"{upa}.meta")
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
 
-        # cfn points to the assembly file outside of the container
-        # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
-        # workdir is shared between the container and the external file system
-        # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+        if retrieve_sample:
+            _download_sample_data(conf, upa)
```

Comment:
I'm thinking we should add a key to the meta file with the sample file as the value, and skip the sample value if it exists. Is there a reason not to do this?

Reply:
👍

Reply:
The sample key constant probably needs to be in a common area so the parser can use it

Reply:
👍

Reply:
:+1: